### PR TITLE
Feat/cypher integration

### PR DESCRIPTION
**Pull request by McSam94**

### CODE REVIEW

File: ts-client/src/vault/strategy/cypher.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+} from '@mercurial-finance/cypher-client';
+import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import { TokenInfo } from '@solana/spl-token-registry';
+import BN from 'bn.js';
+
+import { SEEDS } from '../constants';
+import { AffiliateVaultProgram, VaultProgram } from '../types';
+import { Strategy, StrategyHandler } from '.';
+
+const PROGRAM_ID = 'CYPH3o83JX6jY6NkbproSpdmQ5VWJtxjfJ5P8veyYVu3';
```
Comment (codewithgun) [inline ts-client/src/vault/strategy/cypher.ts:20]: Can we get it from https://github.com/chugach-foundation/cypher-client-ts-v3/blob/main/src/constants/shared.ts#L36?

File: ts-client/src/vault/strategy/cypher.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+const CACHE = new PublicKey('6x5U4c41tfUYGEbTXofFiHcfyx3rqJZsT4emrLisNGGL');
```
Comment (codewithgun) [inline ts-client/src/vault/strategy/cypher.ts:21]: ^Ditto https://github.com/chugach-foundation/cypher-client-ts-v3/blob/main/src/constants/shared.ts#L43

File: ts-client/src/vault/strategy/cypher.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    for (const account of accounts) {
+      remainingAccounts.push({
+        pubkey: account.pubkey,
+        isWritable: !!account.isWritable,
+        isSigner: false,
+      });
+    }
+
+    // prevent duplicate as spot market account pubkey will be add on program side
+    const remainingAccountsWithoutReserve = remainingAccounts.filter(
```
Comment (codewithgun) [inline ts-client/src/vault/strategy/cypher.ts:96]: Why do we need this ?

Comment (00xSam) [reply ts-client/src/vault/strategy/cypher.ts]: Sorry, this can be remove.

File: ts-client/src/vault/strategy/index.ts [hunk 4]
```diff
@@ -55,14 +57,13 @@ export type ReserveState = {
...
   withdraw(
-    tokenInfo: TokenInfo,
     walletPubKey: PublicKey,
     program: VaultProgram,
     strategy: any,
     vault: PublicKey,
     tokenVault: PublicKey,
-    feeVault: PublicKey,
-    lpMint: PublicKey,
+    tokenInfo: TokenInfo,
```
Comment (codewithgun) [inline ts-client/src/vault/strategy/index.ts:65]: Why do we need `tokenInfo`, I don't see it was being used by any strategies.

**Review by codewithgun — APPROVED** [review]

**Pull request merged**